File: Java/BinarySearchTrees/Problem9/Binary Tree to BST.java
package BinarySearchTrees.Problem9;


import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;

class Node<T extends Comparable<T>> {
    private T data;
    private Node<T> left;
    private Node<T> right;

    public Node(T data) {
        this.data = data;
        this.left = this.right = null;
    }

    public T getData() {
        return data;
    }

    public void setData(T data) {
        this.data = data;
    }

    public Node<T> getLeft() {
        return left;
    }

    public void setLeft(Node<T> left) {
        this.left = left;
    }

    public Node<T> getRight() {
        return right;
    }

    public void setRight(Node<T> right) {
        this.right = right;
    }
}


class Solution {
    public static void main(String[] args) {
        // Example 1
        Node<Integer> n1 = new Node<>(1);
        Node<Integer> n2 = new Node<>(2);
        Node<Integer> n3 = new Node<>(3);
        Node<Integer> n4 = new Node<>(4);
        n1.setLeft(n2);
        n2.setLeft(n4);
        n1.setRight(n3);
        System.out.println(getInorder(n1));
        convertToBst(n1);
        System.out.println(getInorder(n1));
    }

    public static <T extends Comparable<T>> void convertToBst(Node<T> root) {
        /**
         * Time complexity is O(n * log(n)) and space complexity is O(n).
         */

        // get the inorder of this binary tree in O(n) time and O(n) space.
        List<T> inorder = getInorder(root);

        // sort the inorder list in O(n * log(n)) time.
        inorder.sort(Comparator.naturalOrder());

        // begin the BST construction in O(n) time and O(n) space.
        finalizeConstruction(root, inorder, Arrays.asList(0));
    }

    private static <T extends Comparable<T>> void finalizeConstruction(Node<T> root, List<T> sortedInorder, List<Integer> i) {
        if (root != null) {
            finalizeConstruction(root.getLeft(), sortedInorder, i);

            // update the root's data with ith pointer in sorted inorder list.
            root.setData(sortedInorder.get(i.getFirst()));
            i.set(0, i.getFirst() + 1);

            finalizeConstruction(root.getRight(), sortedInorder, i);
        }
    }

    private static <T extends Comparable<T>> List<T> getInorder(Node<T> root) {
        List<T> inorder = new ArrayList<>();
        getInorderHelper(root, inorder);
        return inorder;
    }

    private static <T extends Comparable<T>> void getInorderHelper(Node<T> root, List<T> inorder) {
        if (root != null) {
            getInorderHelper(root.getLeft(), inorder);
            inorder.add(root.getData());
            getInorderHelper(root.getRight(), inorder);
        }
    }
}

File: Python/Practice Set 1/Binary Search Tree/Binary Tree to BST.py
class Node:
    def __init__(self, data):
        self.data = data
        self.left = self.right = None


class Solution:
    @staticmethod
    def show_tree(root: Node):
        Solution._show(root)
        print()

    @staticmethod
    def _show(root: Node):
        if root:
            Solution._show(root.left)
            print(root.data, end=" ")
            Solution._show(root.right)

    @staticmethod
    def convert_to_bst(root: Node):
        """
            Time complexity is O(n * log(n)) and space complexity is O(n).
        """

        # get the nodes of the tree in inorder fashion in O(n) time and O(n) space.
        inorder = []
        Solution._get_inorder(root, inorder)

        # sort the inorder nodes data in O(n * log(n)) time.
        sorted_inorder = [i.data for i in inorder]
        sorted_inorder.sort()

        # now assign the inorder nodes with data from sorted inorder list in O(n) time.
        for i in range(len(inorder)):
            inorder[i].data = sorted_inorder[i]

    @staticmethod
    def _get_inorder(root: Node, inorder):
        if root:
            Solution._get_inorder(root.left, inorder)
            inorder.append(root)
            Solution._get_inorder(root.right, inorder)


# Example 1
n1, n2, n3 = Node(1), Node(2), Node(3)
n1.left = n2
n1.right = n3
Solution.convert_to_bst(n1)
Solution.show_tree(n1)

# Example 2
n1, n2, n3, n4 = Node(1), Node(2), Node(3), Node(4)
n1.left = n2
n2.left = n4
n1.right = n3
Solution.convert_to_bst(n1)
Solution.show_tree(n1)

# Example 3
n10, n2, n7, n8, n4 = Node(10), Node(2), Node(7), Node(8), Node(4)
n10.left = n2
n10.right = n7
n2.left = n8
n2.right = n4
Solution.convert_to_bst(n10)
Solution.show_tree(n10)
